compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs

```diff
@@ -87,61 +85,33 @@ createProjectPackageDb opts thisSdkVer deps0 dataDeps = do
     let dalfsFromDars =
             [ ( dropExtension $ takeFileName $ ZipArchive.eRelativePath e
-              , BSL.toStrict $ ZipArchive.fromEntry e)
-            | e <- dalfs
+              , BSL.toStrict $ ZipArchive.fromEntry e
+              )
+            | e <- concatMap edDalfs dars
             ]
+    -- These are dalfs that have been passed in directly as DALFs via data-dependencies.
     dalfsFromFps <-
         forM fpDalfs $ \fp -> do
-            bs <- B.readFile fp
+            bs <- BS.readFile fp
             pure (dropExtension $ takeFileName fp, bs)
     let allDalfs = dalfsFromDars ++ dalfsFromFps
-    pkgs <-
-        forM allDalfs $ \(name, dalf) -> do
-            (pkgId, package) <-
-                either (fail . DA.Pretty.renderPretty) pure $
-                Archive.decodeArchive Archive.DecodeAsMain dalf
-            pure
-                ( pkgId
-                , package
-                , dalf
-                , stringToUnitId $ parseUnitId name pkgId)
-    -- mapping from package id's to unit id's. if the same package is imported with
-    -- different unit id's, we would loose a unit id here.
-    let pkgMap =
-            MS.fromList
-                [(pkgId, unitId) | (pkgId, _pkg, _bs, unitId) <- pkgs]
-    -- order the packages in topological order
-    let (depGraph, vertexToNode, _keyToVertex) =
-            graphFromEdges $ do
-                (pkgId, dalf, bs, unitId) <- pkgs
-                let pkgRefs =
-                        [ pid
-                        | LF.PRImport pid <- toListOf packageRefs dalf
-                        ]
-                let getUid = getUnitId unitId pkgMap
-                let src = generateSrcPkgFromLf getUid (Just "Sdk") dalf
-                let templInstSrc =
-                        generateTemplateInstancesPkgFromLf
-                            getUid
-                            (Just "Sdk")
-                            pkgId
-                            dalf
-                pure
-                    ( (src, templInstSrc, unitId, dalf, bs)
-                    , pkgId
-                    , pkgRefs)
-    let pkgIdsTopoSorted = reverse $ topSort depGraph
+    pkgs <- forM allDalfs $ \(name, dalf) -> do
+        (pkgId, package) <-
+            either (fail . DA.Pretty.renderPretty) pure $
+            Archive.decodeArchive Archive.DecodeAsMain dalf
+        pure (pkgId, package, dalf, stringToUnitId $ parseUnitId name pkgId)
+
     dbPathAbs <- makeAbsolute dbPath
     projectPackageDatabaseAbs <- makeAbsolute projectPackageDatabase
-    forM_ pkgIdsTopoSorted $ \vertex -> do
-        let ((src, templInstSrc, uid, dalf, bs), pkgId, _) =
-                vertexToNode vertex
-        let unitIdStr = unitIdString uid
+
+    let (depGraph, vertexToNode) = buildLfPackageGraph pkgs
+    -- Iterate over the dependency graph in topological order.
+    forM_ (reverse $ topSort depGraph) $ \vertex -> do
```

Comment:
What's the logic behind `reverse $ topSort depGraph` here, specifically why `reverse`? It might be good to add a comment.

Reply:
I’ve added a comment. Basically `topSort` on the original graph would result in traversing a package before its dependencies whereas we want the opposite. I’ve also replaced `reverse . topSort` by `topSort . transposeG` to make this a bit more explicit.

Reply:
👍 thanks!